src/libraries/Microsoft.Extensions.DependencyInjection/ref/Microsoft.Extensions.DependencyInjection.csproj

```diff
@@ -1,6 +1,6 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0;netstandard2.1;net461</TargetFrameworks>
+    <TargetFrameworks>$(NetCoreAppCurrent);$(NetCoreAppMinimum);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
```

Comment:
You are dropping the `netstandard2.1` asset here. We shouldn't be doing that.

Do you have a reason why you are dropping it?

Reply:
Yes, the `netstandard2.1` asset was necessary to be present as it was chosen by .NETCoreApp consumers because a .NETCoreApp asset wasn't available. Now that a .NETCoreApp asset is available, customers are covered by either the .NETFramework, the .NETCoreApp asset or the `netstandard2.0` asset. For remaining consumers who don't support .NETCoreApp, nor .NETFramework, the netstandard2.0 asset is still present.

Reply:
There are other platforms that support `netstandard2.1`, such as Xamarin and Mono. By dropping the `netstandard2.1` assets, you will be degrading the performance on those platforms, because now they will be using the `netstandard2.0` build.

Reply:
Correct but these platforms can continue to use the netstandard2.0 asset (even if it comes with a perf penalty). We don't usually offer netstandard2.1 assets to optimize for these platforms. Also this change targets NET7. At the time this package ships, Xamarin and Mono supports .NETCoreApp.

Reply:
> We don't usually offer netstandard2.1 assets to optimize for these platforms.

Where did you get this information? That is not my understanding.

> these platforms can continue to use the netstandard2.0 asset

Not for libraries where `netstandard2.1` has new APIs - e.g. System.Threading.Channels.

> At the time this package ships, Xamarin and Mono supports .NETCoreApp.

There is also Unity.

Also, it isn't like the day Xamarin supports .NETCoreApp everyone building Xamarin apps will switch to it. There is a transition period while people upgrade.

Reply:
> Where did you get this information? That is not my understanding.

We usually optimize for .NETCoreApp and NetCoreAppCurrent even if that means that we bloat the compile output when targeting other TFM families like .NETFramework and .NETStandard. One example for that is are PlatformAttributes which are included in all tfms (even for .NETFramework which only targets Windows) to avoid additional ifdefs: https://github.com/dotnet/runtime/pull/52582#issuecomment-837725637.

> Not for libraries where netstandard2.1 has new APIs - e.g. System.Threading.Channels.

I agree that for libraries that depend on newer API which isn't available on netstandard2.0 it makes sense to offer a nestandard2.1 asset **as long as we have customers which don't support .NETCoreApp**.

Reply:
> even if that means that we bloat the compile output

This change isn't "bloating the compile output". This change is dropping optimizations that happen at runtime. One of the biggest changes between `netstandard2.0` and `netstandard2.1` is that `2.1` adds all the Span overloads to built-in types - like Stream, String, etc. But there are many other new APIs these libraries are using when targeting `ns2.1`

Further, some of these libraries have drastically different implementations, for example the one I am commenting on - DependencyInjection. The `netstandard2.0` build doesn't use Reflection.Emit, but instead uses System.Linq.Expressions to generate code. This has a very different perf impact than the `netstandard2.1` build, which does use Reflection.Emit.

Reply:
But where do we draw the line? Presumably there are more libraries that would benefit of optimizations that are possible to use/consume in netstandard2.1 but not in netstandard2.0. Based on the low number of netstandard2.1 tfms in our libraries, my assumption was that we don't care much about that scenario.

Reply:
> But where do we draw the line?

My thinking is - if someone has already taken the time to include a `netstandard2.1` build in their package, the line was already drawn. Undoing it here is a take back.

For libraries that don't already take advantage of `ns2.1`, it must not be important to them. So there would be no need to go seeking these out.

Reply:
Makes sense to me. Thanks for the feedback, that was a valuable discussion.